File: R/maxpact.R
maxpact <- function (dist,size,alphac=0.01,mean=FALSE)
{
    if (class(dist) != 'dist') stop('The first argument must be an object of class dist')
    sim <- 1 - as.matrix(dist)
    numplt <- nrow(sim)
    if (mean) morf <- 1
    else morf <- 0
    musuby <- matrix(0,nrow=numplt,ncol=numplt)
    membry <- matrix(0,nrow=numplt,ncol=numplt)
    numset <- 0
    used <- rep(0,numplt)
    musubx <- matrix(0,nrow=numplt,ncol=numplt)
    membrx <- matrix(0,nrow=numplt,ncol=numplt)
    mnsimi <- rep(0,numplt)

    tmp <- .Fortran("maxpact",
                as.double(sim),
                as.integer(numplt),
                as.integer(size),
                as.double(alphac),
                as.integer(morf),
                musuby=as.double(musuby),
                membry=as.integer(membry),
                numset=as.integer(numset),
                as.integer(used),
                as.double(musubx),
                as.integer(membrx),
                as.double(mnsimi),
                PACKAGE='optpart')
    member <- matrix(tmp$membry,nrow=numplt)
    member <- member[1:tmp$numset,1:size]
    musuby <- matrix(tmp$musuby,nrow=numplt)
    musubx <- musuby[1:tmp$numset,1:size]
    distname <- deparse(substitute(dist))
    out <- list(musubx=musubx,member=member,numset=tmp$numset, size=size, distname=distname)
    class(out) <- "mps"
    return(out)
}

mps.test <- function(mps,env,main=deparse(substitute(env)),...)
{
    if (class(mps) != "mps") {
        stop("You must pass an object of class mps from maxpact()")
    }
    if (!is.numeric(env)) {
        stop("You must pass only numeric vectors as environment variables")
    }
    size <- ncol(mps$member)
    nset <- mps$numset
    null <- rep(0,nset)
    for (i in 1:nset) {
        tmp <- sample(1:length(env),size)
        nullmin <- min(env[tmp])
        nullmax <- max(env[tmp])
        null[i] <- nullmax - nullmin
    }
        
    low <- apply(mps$member,1,function(x){min(env[x])})
    high <- apply(mps$member,1,function(x){max(env[x])})
    observed <- high - low
    plot(sort(null),ylim=c(0,max(null)),main=main,ylab="Within-Set Difference")
    points(sort(observed),col=2)
    readline("Hit return\n")
    boxplot(null,observed,names=c("null","observed"),ylab="Within-Set Difference",main=main)
    print(wilcox.test(null,observed))
}

plot.mps <- function(x,...)
{
    plot(x$musubx[1,],ylim=c(0,1),xlab="Size",ylab="Similarity",type="n")
    for (i in 1:x$numset) {
        lines(x$musubx[i,])
    }
}

File: src/maxpact.f90
      subroutine maxpact(sim,numplt,maxsiz,alphac,morf,     &
                 musuby,membry,numset,used,musubx,membrx,   &
                 mnsimi,maxpnt)
 
! passed in
 
      integer numplt
      integer maxsiz
      double precision sim(numplt,numplt)
      double precision alphac
      integer morf
 
! passed out
 
      double precision musuby(numplt,maxsiz)
      integer membry(numplt,maxsiz)
      integer numset
 
! scratch 
 
      integer used(numplt)
      integer maxpnt(numplt)
      integer membrx(numplt,numplt)
      double precision musubx(numplt,numplt)
      double precision mnsimi(numplt)

! local

      double precision maxsim
      double precision temsim
      double precision tmp
      double precision unifrnd
      integer numpnt
      integer flag
      double precision choice
      integer i,j,k
 
!* maxpact ********************* one *********************
      
      call rndstart()
 
      do i=1,numplt
        used = (/ (0,j=1,numplt) /)
        membrx(i,:) = (/ (0,j=1,numplt) /)
        musubx(i,:) = (/ (0,j=1,numplt) /)

        maxpnt(i) = 0
        used(i) = 1
        musubx(i,i) = 1.0
        mnsimi(1) = 0.0
        membrx(i,1) = i
 
        do j=2,maxsiz
          maxsim = 0.0
          numpnt = 0
          do k=1,numplt
            if (used(k) .eq. 1) cycle
            if (morf .eq. 1) then
              temsim = 0.0
              do l=1,numplt
                if (used(l) .eq. 1) then
                  temsim = temsim + sim(k,l)
                endif
              end do   
            else
              temsim = 1.0
              do l=1,numplt
                if (used(l) .eq. 1) then
                  temsim = min(temsim,sim(k,l))
                endif
              end do   
            endif
            if (temsim .gt. maxsim) then
              numpnt = 1
              maxpnt(1) = k
              maxsim = temsim
            else if ((temsim .eq. maxsim) .and. (maxsim .gt. 0.0) ) then
              numpnt = numpnt + 1
              maxpnt(numpnt) = k
            endif 
          end do    
 
          if (numpnt .gt. 1) then
            choice = 0.0
            do l=1,numpnt
              tmp = unifrnd()
              if (tmp .gt. choice) then
                choice = tmp
                maxpnt(1) = maxpnt(l)
              endif
            end do    
          endif
 
          used(maxpnt(1)) = 1
          if (morf .eq. 1) then
            mnsimi(j) = (((j-1)**2-(j-1))/2 * mnsimi(j-1) + maxsim) / ((j**2-j)/2)
          else
            mnsimi(j) = maxsim
          endif
          musubx(i,maxpnt(1)) = mnsimi(j)
          membrx(i,j) = maxpnt(1)
        end do   
      end do   
!
      call rndend()
!
!* maxpact ********************** two ********************************
!
      numset = 0
      do i=1,numplt
        flag = 0
        do 21 j=1,i-1
          do k=1,numplt
            if (musubx(i,k) .ge. alphac .and.             &
                musubx(j,k) .lt. alphac) then
              goto 21
            else if (musubx(i,k) .lt. alphac .and.        &
                     musubx(j,k) .ge. alphac) then
              goto 21
            endif
          end do    
          flag = 1
   21   end do 
        if (flag .eq. 0) then
          numset = numset + 1
          do j=1,maxsiz
            musuby(numset,j) = musubx(i,membrx(i,j))
            membry(numset,j) = membrx(i,j)
          end do    
        endif
      end do   
!
      return
!
      end
